src/view_storage.rs

```diff
@@ -52,18 +52,18 @@ impl ViewStorage {
         }
     }
 
+    /// Returns the deepest view ID encountered traversing parents.  It does *not* guarantee
+    /// that it is a real window root; any caller should perform the same test
+    /// of `window_tracking::is_known_root()` that `ViewId.root()` does before
+    /// assuming the returned value is really a window root.
     pub(crate) fn root_view_id(&self, id: ViewId) -> Option<ViewId> {
-        let mut parent = self.parent.get(id).cloned().flatten();
-        while let Some(parent_id) = parent {
-            match self.parent.get(parent_id).cloned().flatten() {
-                Some(id) => {
-                    parent = Some(id);
-                }
-                None => {
-                    return parent;
-                }
-            }
+        // Rewritten from the original looping version - unless we anticipate view ids
+        // deeper than the possible stack depth, this should be more efficient and
+        // require less cloning.
+        if let Some(p) = self.parent.get(id).unwrap_or(&None) {
+            self.root_view_id(*p)
+        } else {
+            Some(id)
```

Comment:
Even though it's unlikely I don't think we are better off using recursion here. That will place a (unlikely) limit on the max depth. 

Cloning the ViewId should be find. It's only 8 bytes and copies should just all be kept in registers.

Reply:
Happy to leave it as-is too.  Though I think if you have enough view ids to exhaust the stack, you might have some UI usability problems far worse than hitting that limit :-)